docs/dependency_confusion.md

```diff
@@ -0,0 +1,67 @@
+
+To protect yourself even further, use pip’s [hash-checking mode](https://pip.pypa.io/en/stable/reference/pip_install/#hash-checking-mode). Note that pip does not support hash checking for VCS dependencies, e.g. git. Consider transforming your git dependencies to plain https dependencies. For example, if the repository is hosted on github, you can use https://github.com/{org_name}/{repo_name}/{commit_id}.tar.gz to get the tarball for a specific commit.
+
+## npm
+
+*TL;DR: do not use unofficial registries. Even if you try to do so via .npmrc, Cachito will ignore it.*
+
+Npm packages follow the typical package file + lock file approach. The package file is [package.json](https://docs.npmjs.com/cli/v6/configuring-npm/package-json), the lock file is [package-lock.json](https://docs.npmjs.com/cli/v6/configuring-npm/package-lock-json) or [npm-shrinkwrap.json](https://docs.npmjs.com/cli/v6/configuring-npm/shrinkwrap-json). Cachito requires the lock file.
+
+The lock file pins all dependencies to exact versions. For https dependencies, which are impossible to pin, Cachito requires the integrity value (a checksum). For dependencies from the npm registry, Cachito does not require integrity values, but newer versions of npm will always include them. Check if all your dependencies have an integrity value, update your lock file if not.
+
+Do not try to use private registries. If you point npm to a private registry (or any registry other than the official one) via [.npmrc](https://docs.npmjs.com/cli/v6/configuring-npm/npmrc), Cachito will ignore it and look in the official registry anyway.
+
+## yarn
+
+*TL;DR: same as npm, but the handling of unofficial registries is slightly less surprising. Still, you probably should not use them.*
+
+Yarn packages are identical to npm packages except that they use [yarn.lock](https://classic.yarnpkg.com/en/docs/yarn-lock/) as the lock file. Everything that applies to npm applies to yarn.
+
+The one difference is the handling of unofficial registries. If you point yarn to an unofficial registry via .npmrc or [.yarnrc](https://classic.yarnpkg.com/en/docs/yarnrc), this will be reflected in the resolved url in the lock file. Cachito will see that the url does not point to the official registry and will treat the dependency as a plain https dependency. If the url is accessible to Cachito, it will download the dependency directly without relying on npm/yarn dependency resolution. That does not necessarily make using unofficial registries a good idea. If the registry is private, your build will either fail or leak internal package names.
```

Comment:
How to understand "leak internal package names"?

Reply:
Let's say you have something like an `internal-yarn-registry.my-company.com`. In there, you have a private package, `some-private-package`, and you're using this as a dependency for a Cachito request. If Cachito is on the same private network as `internal-yarn-registry.my-company.com`, then it may successfully download `some-private-package`, which will then be present in the request tarball, the request JSON and the content manifest. Depending on what you then do with the finished request (e.g. build a container and a source container), the internal package may end up getting leaked to the outside world.